packages/flutter_tools/lib/src/build_system/targets/font_subset.dart

```diff
@@ -0,0 +1,359 @@
+    final List<Map<String, dynamic>> fontList = _getList(
+      json.decode(fontManifestData),
+      'FontManifest.json invalid: expected top level to be a list of objects.',
+    );
+
+    for (final Map<String, dynamic> map in fontList) {
+      if (map['family'] is! String) {
+        throw FontSubsetException._(
+          'FontManifest.json invalid: expected the family value to be a string, '
+          'got: ${map['family']}.');
+      }
+      final String familyKey = map['family'] as String;
+      if (!families.contains(familyKey)) {
+        continue;
+      }
+      final List<Map<String, dynamic>> fonts = _getList(
+        map['fonts'],
+        'FontManifest.json invalid: expected "fonts" to be a list of objects.',
+      );
+      if (fonts.length != 1) {
+        throw FontSubsetException._(
+          'This tool cannot process icon fonts with multiple fonts in a '
+          'single family.');
+      }
+      if (fonts.first['asset'] is! String) {
+        throw FontSubsetException._(
+          'FontManifest.json invalid: expected "asset" value to be a string, '
+          'got: ${map['assets']}.');
+      }
+      result[familyKey] = fonts.first['asset'] as String;
+    }
+    return result;
+  }
+
+  Future<Map<String, List<int>>> _findConstants(
+    File dart,
+    File constFinder,
+    File appDill,
+  ) async {
+    final List<String> cmd = <String>[
+      dart.path,
+      constFinder.path,
+      '--kernel-file', appDill.path,
+      '--class-library-uri', 'package:flutter/src/widgets/icon_data.dart',
+      '--class-name', 'IconData',
+    ];
+    _logger.printTrace('Running command: ${cmd.join(' ')}');
+    final ProcessResult constFinderProcessResult = await _processManager.run(cmd);
+
+    if (constFinderProcessResult.exitCode != 0) {
+      throwToolExit('ConstFinder failure: ${constFinderProcessResult.stderr}');
+    }
+    final dynamic jsonDecode = json.decode(constFinderProcessResult.stdout as String);
+    if (jsonDecode is! Map<String, dynamic>) {
+      throw FontSubsetException._(
+        'Invalid ConstFinder output: expected a top level JavaScript object, '
+        'got $jsonDecode.');
+    }
+    final Map<String, dynamic> constFinderMap = jsonDecode as Map<String, dynamic>;
+    final _ConstFinderResult constFinderResult = _ConstFinderResult(constFinderMap);
+    if (constFinderResult.hasNonConstantLocations) {
+      _logger.printError('This application cannot tree shake icons fonts. '
+                         'It has non-constant instances of IconData at the '
+                         'following locations:', emphasis: true);
+      for (final Map<String, dynamic> location in constFinderResult.nonConstantLocations) {
+        _logger.printError(
+          '- ${location['file']}:${location['line']}:${location['column']}',
+          indent: 2,
+          hangingIndent: 4,
+        );
+      }
+      throwToolExit('Avoid non-constant invocations of IconData or try to '
```

Comment:
A bit confused by this. Is the font tree shaking on by default? If not, wouldn't the corrective action here be to remove the --tree-shake-icons flag? Instead of throwing a tool exit, can we just print the errors, but then still finishing building the app?

Reply:
This is currently off by default.  At some point I think we'd like to have it on by default. This makes it so if or when we flip that bit, we don't have to worry about updating this exception.

The idea is that if a user has requested to tree shake icons, but their application is unsuitable for treeshaking icons, they should get an error and either fix their app or clearly rebuild without the flag.  Right now, a user would have to opt in to see this error, and the corrective action would likely be to read the output of const_finder and remove any instances of `IconData` that are not const.

Reply:
Other than this, I think I've applied all of your suggestions